spec/java/util/ArrayList.lsl

```diff
@@ -0,0 +1,1205 @@
+        }
+    }
+
+
+    @private
+    @throws(["java.io.IOException", "java.lang.ClassNotFoundException"])
+    fun readObject (s: ObjectInputStream): void
+    {
+        // #problem
+        //Here is cycle and some operations with ObjectInputStream.
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun listIterator (index: int): ListIterator
+    {
+        _rangeCheckForAdd(index, length);
+
+        result = new ListItr(state=Created,
+            cursor=index,
+            expectedModCount=modCount);
+    }
+
+
+    fun listIterator (): ListIterator
+    {
+        result = new ListItr(state=Created,
+            cursor=0,
+            expectedModCount=modCount);
+    }
+
+
+    fun iterator (): Iterator
+    {
+        result = new Itr(state=Created,
+            expectedModCount=modCount);
+    }
+
+
+    fun subList (fromIndex: int, toIndex: int): List
+    {
+        _subListRangeCheck(fromIndex, toIndex, length);
+
+        // #problem
+        //We don't have decision about sublists.
+        //TODO
+        result = new SubList(state=Created,
+            startIndex=fromIndex,
+            endIndex=toIndex);
+    }
+
+
+    fun forEach (action: Consumer): void
+    {
+        // #problem
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun spliterator (): Spliterator
+    {
+        result = new ArrayListSpliterator(state=Initialized,
+            //This is right ? "parent=self"
+            parent=self,
+            origin = 0,
+            est=-1,
+            expectedModCount=0);
+    }
+
+
+    fun removeIf (filter: Predicate): boolean
+    {
+        // #problem
+        //We don't know how to work with lambda.
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun replaceAll (operator: UnaryOperator): void
+    {
+        // #problem
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun sort (c: Comparator): void
+    {
+        var expectedModCount = modCount;
+
+        // #problem
+        //We don't know what to do with sorting of the List.
+        action NOT_IMPLEMENTED();
+
+        if (modCount != expectedModCount)
+        {
+            action THROW_NEW("java.util.ConcurrentModificationException", []);
+        }
+
+        modCount = modCount + 1;
+    }
+}
+
+
+
+@packagePrivate
+@implements("java.util.Iterator")
+@WrapperMeta(
+    src="java.util.ArrayList$Itr",
+    //Maybe will be another name of the dst class
+    dst="ru.spbpu.libsl.overrides.collections.ArrayList_Itr",
+    forceMatchInterfaces=true)
+automaton Itr: int (
+    var cursor: int,
+    var lastRet: int = -1,
+    var expectedModCount: int
+) {
+
+    initstate Initialized;
+
+    shift Allocated -> Initialized by [
+        Itr()
+    ];
+
+    shift Initialized -> self by [
+        // read operations
+        hasNext,
+
+        // write operations
+        next,
+        remove,
+        forEachRemaining
+    ]
+
+
+    // constructors
+
+    constructor Itr()
```

Comment:
It is not possible (except reflection) to call the constructors of this class outside of the `ArrayList`.